### PR TITLE
CLVP Fixes

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

This PR fixes one problem in the rotary embedding implementation, adds an argument to the `ClvpModelForConditionalGeneration.generate` and adds necessary fixes to the test logits. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 @ArthurZucker, @amyeroberts

### CODE REVIEW

File: src/transformers/models/clvp/modeling_clvp.py [hunk 1]
```diff
@@ -312,13 +312,18 @@ def forward(
...
                 key_states[..., rotary_emb_dim:],
             )
+            value_rot, value_pass = (
+                value_states[..., :rotary_emb_dim],
+                value_states[..., rotary_emb_dim:],
+            )
 
             cos, sin = rotary_pos_emb.cos().squeeze(0), rotary_pos_emb.sin().squeeze(0)
-            query_rot, key_rot = apply_rotary_pos_emb(query_rot, key_rot, cos, sin, position_ids)
+            query_rot, key_rot, value_rot = apply_rotary_pos_emb(query_rot, key_rot, value_rot, cos, sin, position_ids)
```
Comment (susnato) [inline src/transformers/models/clvp/modeling_clvp.py:321]: For CLVP we apply the rotary embedding to the value states too, [official repo link](https://github.com/neonbjb/tortoise-tts/blob/80f89987a5abda5e2b082618cd74f9c7411141dc/tortoise/models/xtransformers.py#L629). 
Sorry, I missed it.

I have modified the `test_full_model_integration` to include the check for this change.

File: src/transformers/models/clvp/modeling_clvp.py [hunk 2]
```diff
@@ -1870,6 +1874,13 @@ def generate(
         generation_config.validate()
         self._validate_model_kwargs(model_kwargs.copy())
 
+        # pad input_ids as specified in the original repo
+        # link: https://github.com/neonbjb/tortoise-tts/blob/80f89987a5abda5e2b082618cd74f9c7411141dc/tortoise/api.py#L380
+        input_ids = torch.nn.functional.pad(input_ids, (0, 1))
+        attention_mask = (
+            torch.nn.functional.pad(attention_mask, pad=(0, 1), value=1) if attention_mask is not None else None
+        )
+
```
Comment (susnato) [inline src/transformers/models/clvp/modeling_clvp.py:1883]: One question: Is it ok to keep it here? Or should I move this to the tokenizer file since it's closely related to the tokenizer?

Comment (amyeroberts) [reply src/transformers/models/clvp/modeling_clvp.py:1883]: I would keep this here: this is a modelling specific modification which always pads and only when generating. For the tokenizers, we expect them to behave in a standard way i.e. `padding=False` will not pad the inputs and `padding=True` will pad to the max sequence length.  cc @ArthurZucker WDYT?

Two general comments: 
* For the attention masks, 0 indicates tokens to ignore, and 1 those to attend. If sequences are padded, then we can end up with unusual attentions masks like: `[1, 1, 1, 0, 0, 0, 0, 0, 1]`. Effectively we will always attend to the last token `""`. I'm surprised this is done - regardless of the length of the input sentence we're always adding a token which will have a positional encoding of the largest possible sequence length. Do we know why the original researchers did this? 
*  Won't this break other assumptions in the model? At the moment doing a forward pass, we can pass in either a batch of sequences, all padded to the max sequence length in that batch. or all padded to the max possible sequence length of the model. If the latter is passed, this would make the input_ids 1 token longer than they can be.

File: src/transformers/models/clvp/modeling_clvp.py [hunk 3]
```diff
@@ -1511,11 +1515,6 @@ def fix_speech_decoder_output(self, speech_ids: torch.LongTensor) -> torch.LongT
                 This refers to the output of the decoder model.
         """
         decoder_fixing_codes = self.config.decoder_config.decoder_fixing_codes
-        speech_ids = speech_ids[:, 1:]
```
Comment (amyeroberts) [inline src/transformers/models/clvp/modeling_clvp.py:1514]: We can't remove this here - it would change the output of the current model when doing a forward pass

**Review by amyeroberts — COMMENTED** [review]
Thanks for the work on this! 

Two comments about the padding logic and moving the logic which removes the bos_token.

File: src/transformers/models/clvp/modeling_clvp.py (hunk 2 above)
Comment (susnato) [reply src/transformers/models/clvp/modeling_clvp.py:1883]: > Do we know why the original researchers did this? 

No, 

> If sequences are padded, then we can end up with unusual attentions masks like: [1, 1, 1, 0, 0, 0, 0, 0, 1]. Effectively we will always attend to the last token ""

In the official repo, they don't process texts in batches so no `attention_mask` and no issue there :) .

> this would make the input_ids 1 token longer than they can be.

To remove this possibility, instead of adding it here I think this line should be added to the pipeline code in the diffusers library where we check for the length of the sequence before passing it to the model. WDYT?

Comment (amyeroberts) [reply src/transformers/models/clvp/modeling_clvp.py:1883]: > No,

OK - I suggest raising an issue on the repo to ask

> In the official repo, they don't process texts in batches so no attention_mask and no issue there :) .

What we'd need to protect against is people using batches in our repo - how would a user know not to do this? We'd have to do something like raising an exception if they try

> To remove this possibility, instead of adding it here I think this line should be added to the pipeline code in the diffusers library where we check for the length of the sequence before passing it to the model. WDYT?

It really depends on what this code is for (which we need to find out). All modeling logic should be contained within the model itself. I'd expect the sequence length info to be accessible in `self.config`

**Review by ArthurZucker — COMMENTED** [review]
LGTM appart from the regression in the forward regarding the speech ids (unless it should only be done when generating?)

File: src/transformers/models/clvp/modeling_clvp.py (hunk 1 above)
Comment (susnato) [reply src/transformers/models/clvp/modeling_clvp.py:321]: Hi @ArthurZucker, I am sorry for the mistake. I was focused on checking the logits of the Decoder Model (which does not use the rotary embeddings thus the logits were properly validated for that one) and didn't focused much into the CLVP Encoder models.

File: src/transformers/models/clvp/modeling_clvp.py (hunk 2 above)
Comment (susnato) [reply src/transformers/models/clvp/modeling_clvp.py:1883]: Hi @amyeroberts, I talked with @manmay-nakhashi (the maintainer of the `tortoise-tts` repo) on Slack and he confirmed that instead of adding the `0` at the end, it should be added where the valid `input_ids` end, so that the `0` will be considered as a valid token and we will have usual `attention_mask`. So,

input_ids - `[5, 6, 7, [PAD], [PAD]]`, attention_mask - `[1, 1, 1, 0, 0]`

After adding the `0` - 

input_ids - `[5, 6, 7, 0, [PAD], [PAD]]`, attention_mask - `[1, 1, 1, 1, 0, 0]`

> this would make the input_ids 1 token longer than they can be.

To avoid this I think we should just check for the `self.config.decoder_config.max_text_tokens - 3` and raise an error if the user gives longer tokens than that. The 3 is because we already add bos and eos tokens for the [ClvpConditioningEncoder](https://github.com/huggingface/transformers/blob/f18c95b49cbb675764f0cd19b56fb70f39a62bb2/src/transformers/models/clvp/modeling_clvp.py#L604) and we are adding one more extra token here :)   

WDYT?

File: src/transformers/models/clvp/modeling_clvp.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/clvp/modeling_clvp.py:321]: No worries! 🤗

**Comment by susnato** [comment]
Hi @amyeroberts, @ArthurZucker, I have worked on the comments and added the logic to add `bos` and `eos` tokens according to this [comment](https://github.com/huggingface/transformers/pull/27547#discussion_r1399547566).  

Please let me know if you are Ok with this change.

File: src/transformers/models/clvp/modeling_clvp.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/models/clvp/modeling_clvp.py]: @susnato Thanks for looking into this and writing this details explanation! 

In principle sounds good to me. Could you link to where you think the exception should be raised?

File: src/transformers/models/clvp/modeling_clvp.py [hunk 4]
```diff
@@ -1855,13 +1901,27 @@ def generate(
...
+
+        # If the input sequences are larger than (self.config.decoder_config.max_text_tokens - 3) then raise error, because we add 1 bos tokens and
+        # 2 eos tokens to the input_ids in ClvpConditioningEncoder.
+        sequence_length = input_ids.shape[-1]
+        if sequence_length > (self.config.decoder_config.max_text_tokens - 3):
+            raise ValueError(
+                f"Maximum sequence length reached! Found input_ids of length {sequence_length}."
+                f"Please make sure that the maximum length of input_ids is {self.config.decoder_config.max_text_tokens - 3}"
+            )
+
```
Comment (susnato) [inline src/transformers/models/clvp/modeling_clvp.py:1924]: This is where we should raise the exception IMO. @amyeroberts

File: src/transformers/models/clvp/modeling_clvp.py [hunk 5]
```diff
@@ -1855,13 +1901,27 @@ def generate(
...
             output_hidden_states (`bool`, *optional*):
                 Whether or not to return the hidden states of decoder model, text encoder and speech encoder models.
 
         Returns:
             `ClvpOutput` or tuple: A `ClvpOutput` (if `return_dict_in_generate=True` or when
             `config.return_dict_in_generate=True`) or a tuple.
         """
+
+        # If the input sequences are larger than (self.config.decoder_config.max_text_tokens - 3) then raise error, because we add 1 bos tokens and
+        # 2 eos tokens to the input_ids in ClvpConditioningEncoder.
```
Comment (ArthurZucker) [inline src/transformers/models/clvp/modeling_clvp.py:1917]:
```suggestion
        # If the input sequences are larger than (self.config.decoder_config.max_text_tokens - 3) then raise error, because we need to add 3 tokens ( 1 bos tokens and
        # 2 eos tokens) to the input_ids in ClvpConditioningEncoder to properly sample
```

File: src/transformers/models/clvp/modeling_clvp.py [hunk 6]
```diff
@@ -1855,13 +1901,27 @@ def generate(
                 priority: 1) from the `generation_config.json` model file, if it exists; 2) from the model
                 configuration. Please note that unspecified parameters will inherit [`~generation.GenerationConfig`]'s
                 default values, whose documentation should be checked to parameterize generation.
+            pad_to_max_mel_tokens (`int`, *optional*):
+                Pads generated speech_ids to the specified value. This is to implement the same logic from the official
+                repo, link: https://github.com/neonbjb/tortoise-tts/blob/80f89987a5abda5e2b082618cd74f9c7411141dc/tortoise/api.py#L430
+                and to make sure the logits are same.
```
Comment (ArthurZucker) [inline src/transformers/models/clvp/modeling_clvp.py:1907]: If this is only used to get the exact same logits, but is less efficient we should also say that for faster sampling we recommend not to use it.

File: src/transformers/models/clvp/modeling_clvp.py [hunk 7]
```diff
@@ -107,7 +106,51 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
     sin = sin[position_ids].unsqueeze(unsqueeze_dim)
     q_embed = (q * cos) + (rotate_half(q) * sin)
     k_embed = (k * cos) + (rotate_half(k) * sin)
-    return q_embed, k_embed
+    v_embed = (v * cos) + (rotate_half(v) * sin)
+    return q_embed, k_embed, v_embed
+
+
+def pad_extra_bos_eos_tokens(
```
Comment (ArthurZucker) [inline src/transformers/models/clvp/modeling_clvp.py:113]:
```suggestion
def _pad_extra_bos_eos_tokens(
```

Comment (ArthurZucker) [reply src/transformers/models/clvp/modeling_clvp.py:113]: let's just mark it as internal function

**Review by ArthurZucker — APPROVED** [review]
Thanks 😉 left small nits related to doc mostly

File: src/transformers/models/clvp/modeling_clvp.py (hunk 6 above)
Comment (susnato) [reply src/transformers/models/clvp/modeling_clvp.py:1907]: yes, let's add a line - "This does not affect generation quality much so please don't consider using it since it is less efficient."

**Comment by susnato** [comment]
Hey, it's always a pleasure to contribute! :raised_hands:
